### PR TITLE
[App] Add --zip option to cp command

### PR DESCRIPTION
**Pull request by luca3rd**
## What does this PR do?

- Adds command to download content from the Lightning Filesystem as a zip file.
```
➜ lightning pwd
/default-project

➜ lightning ls
wide-tan-ryq8

➜ lightning cp r:wide-tan-ryq8 . --zip
Downloaded to ./wide-tan-ryq8.zip

➜ unzip -l ./wide-tan-ryq8.zip 
Archive:  ./wide-tan-ryq8.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  02-17-2023 11:34   code/.lightning-app-sync
       41  02-17-2023 11:34   code/app.py
---------                     -------
       41                     2 files
```
- Fixes bug in `cp._sanitize_path`
- Refactors token client
- Bumps `rich` version to support progress bars w/ indefinite size

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda

### CODE REVIEW

**Comment by tchaton** [comment]
@luca3rd This is kinda of a confusion API. 

Won't it be better to have `lightning cp r:. . --zip`, so we don't introduce a new command ?

Best,
T.C

File: src/lightning/app/cli/commands/cp.py [hunk 1]
```diff
@@ -162,6 +174,45 @@ def _upload(source_file: str, presigned_url: ApplyResult, progress: Progress, ta
     file_uploader.upload()
 
 
+def _zip_files(live: Live, remote_src: str, local_dst: str) -> None:
+    live.stop()
```
Comment (tchaton) [inline src/lightning/app/cli/commands/cp.py:178]: Let's move the live right before the progress.start. Everything below as a small delay.

File: src/lightning/app/cli/commands/cp.py [hunk 2]
```diff
@@ -228,7 +279,7 @@ def _sanitize_path(path: str, pwd: str) -> Tuple[str, bool]:
         if path == ".":
             path = pwd
         else:
-            path = os.path.join(pwd, path[1:])
+            path = os.path.join(pwd, path)
```
Comment (tchaton) [inline src/lightning/app/cli/commands/cp.py:282]: I remember there were a reason for it. Not sure.

**Review by tchaton — APPROVED** [review]
Nice work !

File: src/lightning/app/cli/commands/cp.py (hunk 2 above)
Comment (luca3rd) [reply src/lightning/app/cli/commands/cp.py:282]: I added a test for this function, because I was trying to sanitize `r:default-project` from `/` and was getting `/efault-project`.

If the edge case ever pops up again, let's be sure to add to the test.

**Review by nicolai86 — APPROVED** [review]

File: requirements/pytorch/extra.txt [hunk 3]
```diff
@@ -6,5 +6,5 @@ matplotlib>3.1, <3.6.2
 omegaconf>=2.0.5, <2.4.0
 hydra-core>=1.0.5, <1.4.0
 jsonargparse[signatures]>=4.18.0, <4.19.0
-rich>=10.14.0, !=10.15.0.a, <13.0.0
```
Comment (luca3rd) [inline requirements/pytorch/extra.txt:9]: Required to get the `oldest` matrix test to pass (our script changes `>=` to `==` which causes a dependency conflict -> `==10.14.0` v `>12.3.0`)

File: requirements/app/base.txt [hunk 4]
```diff
@@ -19,7 +19,7 @@ Jinja2<=3.1.2
 pydantic<=1.10.4
 PyYAML<=6.0
 requests<=2.28.1
-rich<=13.0.1
+rich<=13.0.1, >12.3.0
```
Comment (Borda) [inline requirements/app/base.txt:22]:
```suggestion
rich >=12.3.0, <=13.0.1
```

File: requirements/pytorch/extra.txt (hunk 3 above)
Comment (Borda) [reply requirements/pytorch/extra.txt:9]: Thinking we can bump the version also here to be consistent..
cc: @awaelchli @carmocca

Comment (luca3rd) [reply requirements/pytorch/extra.txt:9]: I added a note in the changelogs.
